packages/jest-mock/src/__tests__/index.test.ts

```diff
@@ -1604,6 +1604,179 @@ describe('moduleMocker', () => {
       expect(spy2.mock.calls).toHaveLength(1);
     });
   });
+
+  describe('mockProperty', () => {
+    it('should work', () => {
+      const obj = {
+        property: 1,
+      };
+
+      const replaced = moduleMocker.mockProperty(obj, 'property', 2);
+
+      expect(obj.property).toBe(2);
+
+      replaced.mockRestore();
+
+      expect(obj.property).toBe(1);
+    });
+
+    it('should allow mocking a property multiple times', () => {
+      const obj = {
+        property: 1,
+      };
+
+      const replacedFirst = moduleMocker.mockProperty(obj, 'property', 2);
+
+      const replacedSecond = moduleMocker.mockProperty(obj, 'property', 3);
+
+      expect(obj.property).toBe(3);
+
+      replacedSecond.mockRestore();
+
+      expect(obj.property).toBe(1);
+
+      replacedFirst.mockRestore();
+
+      expect(obj.property).toBe(1);
+    });
+
+    describe('should throw', () => {
+      it('when object is not provided', () => {
+        expect(() => {
+          moduleMocker.mockProperty(null, 'property', 1);
+        }).toThrow(
+          'mockProperty could not find an object on which to replace property',
+        );
+      });
+
+      it('when primitive value is provided instead of object', () => {
+        expect(() => {
+          moduleMocker.mockProperty(1, 'property', 1);
+        }).toThrow('Cannot mock property on a primitive value; number given');
+      });
+
+      it('when function is provided instead of object', () => {
+        expect(() => {
+          moduleMocker.mockProperty(() => {}, 'property', 1);
+        }).toThrow('Cannot mock property on a primitive value; function given');
+      });
+
+      it('when property name is not provided', () => {
+        expect(() => {
+          moduleMocker.mockProperty({}, null, 1);
+        }).toThrow('No property name supplied');
+      });
+
+      it('when property is not defined', () => {
+        expect(() => {
+          moduleMocker.mockProperty({}, 'doesNotExist', 1);
+        }).toThrow('doesNotExist property does not exist');
+      });
+
+      it('when property is not configurable', () => {
+        expect(() => {
+          const obj = {};
+
+          Object.defineProperty(obj, 'property', {
+            configurable: false,
+            value: 1,
+            writable: false,
+          });
+
+          moduleMocker.mockProperty(obj, 'property', 2);
+        }).toThrow('property is not declared configurable');
+      });
+
+      it('when mocking with value of different type', () => {
+        expect(() => {
+          moduleMocker.mockProperty({property: 1}, 'property', 'string');
+        }).toThrow(
+          'Cannot mock the property property because it is not a number; string given instead',
```

Comment:
This limitation seems unnecessary. This would basically not allow mocking all types in a   property that has a union type.